src/utility/png.cpp

```diff
@@ -0,0 +1,190 @@
+        foreground, background, out);
+}
+
+bool write_png(std::istream& in, const uint32_t size,
+    std::ostream& out)
+{
+    return write_png(in, size, png::dpi, png::margin,
+        png::inches_per_meter, png::foreground, png::background, out);
+}
+
+extern "C" void internal_png_sink_write(png_structp png_ptr,
+    png_bytep data, png_size_t length)
+{
+    ostream_writer& sink = *reinterpret_cast<ostream_writer*>(
+        png_get_io_ptr(png_ptr));
+    sink.write_data(reinterpret_cast<const uint8_t*>(data), size_t(length));
+}
+
+extern "C" void user_error_fn(png_structp png_ptr, png_const_charp error_msg)
+{
+    throw std::runtime_error(error_msg);
+}
+
+bool write_png(std::istream& in, const uint32_t size,
+    const uint32_t dpi, const uint32_t margin, const uint32_t inches_per_meter,
+    const png::color foreground, const png::color background,
+    std::ostream& out)
+{
+    uint32_t version, width;
+    int32_t x, y, xx, yy, bit;
+    const auto bits_per_byte = 8;
+
+    if (size == 0)
+        return false;
+
+    istream_reader source(in);
+    source.read_data(reinterpret_cast<uint8_t*>(&version), sizeof(uint32_t));
+    source.read_data(reinterpret_cast<uint8_t*>(&width), sizeof(uint32_t));
+
+    data_chunk data = source.read_data(width * width);
+
+    const auto margin_value = 0xff;
+    const auto realwidth = (width + margin * 2) * size;
+    const auto row_size = (realwidth + 7) / bits_per_byte;
+
+    data_chunk row;
+    row.reserve(row_size);
+
+    png_structp png_ptr = png_create_write_struct(
+        PNG_LIBPNG_VER_STRING, NULL, NULL, NULL);
+    if (png_ptr == NULL)
+        return false;
+
+    png_infop info_ptr = png_create_info_struct(png_ptr);
+    if (info_ptr == NULL)
+        return false;
+
+    png_color raw_palette;
+    png_colorp palette = &raw_palette;
+    palette[0].red   = foreground[0];
+    palette[0].green = foreground[1];
+    palette[0].blue  = foreground[2];
+    palette[1].red   = background[0];
+    palette[1].green = background[1];
+    palette[1].blue  = background[2];
+
+    png_byte alpha_values[2];
+    alpha_values[0] = foreground[3];
+    alpha_values[1] = background[3];
+    png_set_PLTE(png_ptr, info_ptr, palette, 2);
+    png_set_tRNS(png_ptr, info_ptr, alpha_values, 2, NULL);
+
+
+    ostream_writer sink(out);
+    png_set_write_fn(png_ptr, &sink, internal_png_sink_write, NULL);
+    png_set_error_fn(png_ptr, NULL, user_error_fn, NULL);
+
+    try
+    {
+        png_set_IHDR(png_ptr, info_ptr,
+                     realwidth, realwidth,
+                     1,
+                     PNG_COLOR_TYPE_PALETTE,
+                     PNG_INTERLACE_NONE,
+                     PNG_COMPRESSION_TYPE_DEFAULT,
+                     PNG_FILTER_TYPE_DEFAULT);
+        png_set_pHYs(png_ptr, info_ptr,
+                     dpi * inches_per_meter,
+                     dpi * inches_per_meter,
+                     PNG_RESOLUTION_METER);
+        png_write_info(png_ptr, info_ptr);
```

Comment:
Style: try to avoid floating stacks, they don't refactor well. Generally we avoid stacks, but if they improve readability use the standard indent. This then amounts to just breaking lines early:
```cpp
    png_set_IHDR(
        png_ptr, info_ptr,
        realwidth, realwidth,
        1,
        PNG_COLOR_TYPE_PALETTE,
        PNG_INTERLACE_NONE,
        PNG_COMPRESSION_TYPE_DEFAULT,
        PNG_FILTER_TYPE_DEFAULT);
    png_set_pHYs(
        png_ptr, info_ptr,
        dpi * inches_per_meter,
        dpi * inches_per_meter,
        PNG_RESOLUTION_METER);
    png_write_info(png_ptr, info_ptr);
```